### ISSUE TITLE
no-unused-declaration false errors with jsx

### ISSUE THREAD

**Issue by Hotell** [cartant/tslint-etc#7]
React, Preact uses `import React form 'react'` / `import {h} from 'preact'` for JSX transpilation so both are used although we got this error:

![image](https://user-images.githubusercontent.com/1223799/44797267-14b78c80-abaf-11e8-9f96-c77abb9b724c.png)

**Comment by cartant** [comment]
Thanks, Martin. I've added a bunch of failing tests for the currently-open issues. Hopefully, I'll get them fixed today.

**Comment by cartant** [comment]
Should be fixed in 1.2.4.

**Comment by Hotell** [comment]
packages versions:

```json
{
  "tslint": "5.11.0",
  "tslint-config-prettier": "1.15.0",
  "tslint-config-standard": "8.0.0",
  "tslint-etc": "1.2.4",
  "tslint-react": "3.6.0",
  "typescript": "2.9.2"
}
```

Still doesn't work. Also I got now lot's of strange exception errors:

![image](https://user-images.githubusercontent.com/1223799/44950084-3333b700-ae3f-11e8-87f7-4740619905b7.png)

![image](https://user-images.githubusercontent.com/1223799/44950098-62e2bf00-ae3f-11e8-84e5-3fe798f52cce.png)
cce.png)

**Comment by cartant** [comment]
Can you give me the smallest single file you have that effects the `Cannot read property 'forEach' of undefined` error? I can see why it's happening - it cannot find any declarations for a symbol - but I'd like to reproduce it in a representative fixture.

And a single, simple file that effects the false positive would be useful, too. FYI, the passing fixture is [here](https://github.com/cartant/tslint-etc/blob/master/fixtures/issues/7-preact/fixture.tsx.lint). And the implementation was a [small change](https://github.com/cartant/tslint-etc/blob/v1.2.4/source/rules/noUnusedDeclarationRule.ts#L105-L111).

**Comment by Hotell** [comment]
Hmm this is kinda old project and has various custom type mapping:

**tslint:**

```json
{
  "extends": [
    "tslint:recommended",
    "tslint-config-standard",
    "tslint-etc",
    "tslint-react",
    "tslint-config-prettier"
  ],
  "rules": {
    // types related rules
    "no-inferrable-types": true,
    "no-namespace": {
      "options": [
        "allow-declarations"
      ]
    },
    "interface-name": {
      "options": "never-prefix"
    },
    // "no-unsafe-any": true,

    // core lint rules
    "ban": {
      "options": [
        {
          "name": "parseInt",
          "message": "use #type-coercion -> Number(val)"
        },
        {
          "name": "parseFloat",
          "message": "use #type-coercion -> Number(val)"
        },
        {
          "name": "Array",
          "message": "use #array-constructor"
        },
        {
          "name": [
            "describe",
            "only"
          ],
          "message": "don't focus spec blocks"
        },
        {
          "name": [
            "it",
            "only"
          ],
          "message": "don't focus tests"
        }
      ]
    },
    "no-magic-numbers": true,
    "no-require-imports": true,
    "no-boolean-literal-compare": true,
    "no-invalid-this": {
      "options": "check-function-in-method"
    },
    "no-invalid-template-strings": true,
    "ordered-imports": true,
    "prefer-template": true,
    "newline-before-return": true,
    // "match-default-export-name": true,
    "no-parameter-reassignment": true,
    // "file-name-casing": {
    //   "options": [
    //     "kebab-case"
    //   ]
    // },
    "switch-default": true,

    // tslint:recommended overrides
    "no-any": true,
    "member-access": {
      "options": [
        "no-public"
      ]
    },
    "object-literal-sort-keys": false,
    "interface-over-type-literal": false,

    // tslint-config-standard overrides
    "no-use-before-declare": false,

    // tslint-etc rules
    // "no-unused-declaration": true,

    // tslint-react rules
    "jsx-boolean-value": {
      "options": "never"
    }
  },
  // apply the same rules for any JS if allowJS is gonna be used
  "jsRules": true
}
```

**tsconfig:**

```json
{
  "compilerOptions": {
    "module": "esnext",
    "target": "es5",
    "moduleResolution": "node",
    "strict": true,
    "strictFunctionTypes": false,
    "suppressImplicitAnyIndexErrors": true,
    "forceConsistentCasingInFileNames": true,
    "allowSyntheticDefaultImports": true,
    "esModuleInterop": true,
    "sourceMap": true,
    "importHelpers": true,
    "jsx": "react",
    "jsxFactory": "h",
    "baseUrl": "./",
    "rootDir": "./",
    "outDir": "ts-output",
    "skipLibCheck": true,
    "typeRoots": [
      "manual_typings",
      "node_modules/@types"
    ],
    "paths": {
      "custom_modules/*": [
        "src/custom_modules/*"
      ],
      "preact": [
        "manual_typings/preact/new"
      ],
      "react": [
        "node_modules/@types/react"
      ],
      "preact-render-to-string/jsx": [
        "manual_typings/preact-render-to-string/jsx"
      ]
    },
    "lib": [
      "dom",
      "es2015",
      "es2016.array.include"
    ]
  },
  "include": [
    "src"
  ],
  "exclude": [
    "node_modules"
  ]
}
```

here is smallest file I found that contains this error:

```tsx
import { h, SFC, SyntheticEvent } from 'preact'

// maybe this is the culprit -> 'custom_modules/util' is defined via path alias within tsconfig
import { identity } from 'custom_modules/util'

type Props = {
  checked?: boolean
  disabled?: boolean
  id?: string
  name?: string
  value?: string
  onChange?(ev: SyntheticEvent<{}>): void
}

const Checkbox: SFC<Props> = ({ checked = false, disabled, id, name, value, onChange = identity }) => {
  return (
    <input
      type="checkbox"
      id={id}
      name={name}
      value={value}
      checked={checked}
      disabled={disabled}
      onChange={onChange}
    />
  )
}

export { Checkbox }
```

hope this helps

**Comment by cartant** [comment]
This should be fixed in 1.2.5, along with #8. It turns out the this was the easy one and the other was the hard(er) one.

**Comment by Hotell** [comment]
everything works like a charm ! many thanks Nicholas !  Cheers dude 🍻

![image](https://user-images.githubusercontent.com/1223799/45204770-87e09300-b280-11e8-9065-50dc1f3610c7.png)

**Comment by cartant** [comment]
No worries.
